### PR TITLE
SmallRye Fault Tolerance 4.0 blog post

### PR DESCRIPTION
**Pull request by Ladicek**

### CODE REVIEW

File: _posts/2020-01-21-fault-tolerance-4-0.adoc [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+= Fault Tolerance 4.0
+
+{uri-smallrye-ft}[SmallRye Fault Tolerance] is our implementation of {uri-microprofile-ft}[Eclipse MicroProfile Fault Tolerance].
+It has been originally based on {uri-hystrix}[Hystrix], the Netflix library for latency and fault tolerance for distributed systems.
+Lately, we realized that Hystrix is not the best fit for implementing MicroProfile Fault Tolerance, because Hystrix meshes all fault tolerance concerns into one, while MicroProfile Fault Tolerance describes mostly a layered architecture.
+The MicroProfile Fault Tolerance specification also requires certain features that Hystrix intentionally doesn't provide; most importantly, the ability to interrupt threads that Hystrix itself didn't create.
+
+We have been talking on and off about removing Hystrix and using our own implementation of core fault tolerance strategies, one that would be better suited as a base for implementing MicroProfile Fault Tolerance, but only a few months ago did we finally start working on it.
+Thanks to heroic work of Michał Szynkiewicz, the new implementation didn't take long to pass the entire MicroProfile Fault Tolerance TCK.
```
Comment (kenfinnigan) [inline _posts/2020-01-21-fault-tolerance-4-0.adoc]: Not critical, but maybe a link to Michal's GH with his name?

File: _posts/2020-01-21-fault-tolerance-4-0.adoc [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+}
+----
+
+The `InvocationContext` is a `Callable` that represents the method invocation guarded by this fault tolerance strategy.
+The fault tolerance strategy does its work around `ctx.call()`.
+For example, it can catch exceptions, invoke `ctx.call()` multiple times, invoke something else, etc.
+
+In the codebase, you can find implementations of this interface for all the strategies required by MicroProfile Fault Tolerance: retry, fallback, timeout, circuit breaker or bulkhead.
+Asynchronous invocation, delegated to a thread pool, is of course also supported.
+When multiple fault tolerance strategies are supposed to be used to guard one method, they can form a chain.
```
Comment (kenfinnigan) [inline _posts/2020-01-21-fault-tolerance-4-0.adoc]: Does it need "can"? Just wondering if "they form a chain" makes it clear they form a chain?

File: _posts/2020-01-21-fault-tolerance-4-0.adoc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+@Retry(...)
+@Timeout(...)
+public void doSomething() {
+    ...
+}
+----
+
+The chain of fault tolerance strategies will look roughly like this:
+
+[source]
```
Comment (kenfinnigan) [inline _posts/2020-01-21-fault-tolerance-4-0.adoc]: Purely for my curiosity, does the ordering of the annotations on the method dictate the chaining?

If so, does that have any implications for the ordering? ie. that a specific order is recommended to chain them in an appropriate manner? Should `Fallback` always be the first annotation so it's the outermost chain?

File: _data/authors.yaml [hunk 4]
```diff
@@ -12,3 +12,10 @@ kenfinnigan:
...
+lthon:
+  name: "Ladislav Thon"
+  email: "ladicek@gmail.com"
+  emailhash: "238e8b2100a5366b226234b946026462"
+  job_title: ""
+  twitter: "_Ladicek"
+  bio: 'Ladislav is a Senior Software Engineer at Red Hat, where he leads the Thorntail project and contributes to MicroProfile. Previously, he lead the Red Hat OpenShift Application Runtimes QE team, and before that spent a few years in the JBoss Enterprise Application Platform QE team. He has been working as a developer for more than a decade, and besides programming, he likes good books and craft beer.'
```
Comment (michalszynkiewicz) [inline _data/authors.yaml]:
```suggestion
  bio: 'Ladislav is a Senior Software Engineer at Red Hat, where he leads the Thorntail project and contributes to MicroProfile. Previously, he led the Red Hat OpenShift Application Runtimes QE team, and before that spent a few years in the JBoss Enterprise Application Platform QE team. He has been working as a developer for more than a decade, and besides programming, he likes good books and craft beer.'
```

File: _posts/2020-01-21-fault-tolerance-4-0.adoc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+For that reason, we would probably mark these prototypes as experimental API, and if they get specified in MicroProfile, we'd rapidly deprecate and remove them.
+
+There's also some implementation aspects that we'd like to finetune.
+
+The metrics collection and configuration handling code mostly comes from the old implementation, and is due for serious refactoring.
+
+As mentioned above, each threadpool-style bulkhead currently gets a fresh thread pool.
+This is quite inefficient, and can be significantly improved.
+
+We want to provide comprehensive debug and trace logs, and we intend to move to JBoss Logging for all of our logging needs.
```
Comment (michalszynkiewicz) [inline _posts/2020-01-21-fault-tolerance-4-0.adoc]: aren't we on jboss logging already?

**Comment by michalszynkiewicz** [comment]
looks good to me, I only found one typo and had one question.
Thanks for mentioning me by name :) Maybe find a way to include yours as the author as well?

File: _posts/2020-01-21-fault-tolerance-4-0.adoc (hunk 3 above)
Comment (michalszynkiewicz) [reply _posts/2020-01-21-fault-tolerance-4-0.adoc]: The order in the chain is specified by MP FT and the order in which the annotations are defined does not modify it.

Comment (kenfinnigan) [reply _posts/2020-01-21-fault-tolerance-4-0.adoc]: awesome, thanks.

Just a coincidence that the annotations matched up then

File: _posts/2020-01-21-fault-tolerance-4-0.adoc (hunk 5 above)
Comment (Ladicek) [reply _posts/2020-01-21-fault-tolerance-4-0.adoc]: Hmm, perhaps so. I'll check again.

File: _data/authors.yaml (hunk 4 above)
Comment (Ladicek) [reply _data/authors.yaml]: Fixed, thanks!

File: _posts/2020-01-21-fault-tolerance-4-0.adoc (hunk 5 above)
Comment (Ladicek) [reply _posts/2020-01-21-fault-tolerance-4-0.adoc]: Fixed, thanks!

File: _posts/2020-01-21-fault-tolerance-4-0.adoc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+:uri-smallrye-ft: https://github.com/smallrye/smallrye-fault-tolerance
+:uri-microprofile-ft: https://github.com/eclipse/microprofile-fault-tolerance
+:uri-microprofile-context-propagation: https://github.com/eclipse/microprofile-context-propagation
+:uri-microprofile-opentracing: https://github.com/eclipse/microprofile-opentracing
+:uri-hystrix: https://github.com/Netflix/Hystrix
+
+= Fault Tolerance 4.0
+
+{uri-smallrye-ft}[SmallRye Fault Tolerance] is our implementation of {uri-microprofile-ft}[Eclipse MicroProfile Fault Tolerance].
+It has been originally based on {uri-hystrix}[Hystrix], the Netflix library for latency and fault tolerance for distributed systems.
```
[the remark this answers is not in the document]
Comment (Ladicek) [reply _posts/2020-01-21-fault-tolerance-4-0.adoc]: Fixed, thanks!

File: _posts/2020-01-21-fault-tolerance-4-0.adoc (hunk 1 above)
Comment (Ladicek) [reply _posts/2020-01-21-fault-tolerance-4-0.adoc]: Fixed, thanks!

File: _posts/2020-01-21-fault-tolerance-4-0.adoc (hunk 2 above)
Comment (Ladicek) [reply _posts/2020-01-21-fault-tolerance-4-0.adoc]: Fixed, thanks!

File: _posts/2020-01-21-fault-tolerance-4-0.adoc (hunk 3 above)
Comment (Ladicek) [reply _posts/2020-01-21-fault-tolerance-4-0.adoc]: I reordered the annotations and added an explanatory sentence. Good point, I didn't even realize that the order was the same :-)

**Comment by Ladicek** [comment]
I have addressed all comments here, and also added 2 small examples to better illustrate how FTStrategies work.

**Comment by kenfinnigan** [comment]
John has asked if you can add a small comment about Netflix Hystrix being in maintenance mode as well. Is that doable?

**Comment by Ladicek** [comment]
Good point, I'll submit a subsequent PR.

**Pull request merged**